Jupyterlab-notify version 2 with enhanced features

Darshan808 opened the pull request:
## Description

This PR proposes version 2 of the `jupyterlab-notify` extension, bringing exciting new features, a refreshed user interface, simplified configuration, and the ability to receive notifications via email and Slack—even when your browser is closed.

## Key Features and Improvements

- **Enhanced UI and Notification Modes**  
  Version 1 depends on legacy IPython cell magic module which requires magic commands to configure notifications. In version 2, you can now set notification modes for each cell effortlessly using the cell toolbar. Choose from:  
  - `default`: Triggers notifications based on a preset threshold.  
  - `never`: Silences notifications for that cell.  
  - `on-error`: Alerts you only if the cell fails.  
  - `custom-timeout`: Notifies you if execution exceeds your specified time limit.  

  ![Cell Toolbar](https://github.com/user-attachments/assets/7de31c42-c480-4a25-a8ee-e1d5cffa7e05)  
  ![Notification Modes](https://github.com/user-attachments/assets/74425c5c-c844-4ec2-b61a-c86962a65b24)

- **Server-Side Execution Support**  
  With optional server-side execution dependencies, notifications keep working even after you close your browser. Perfect for long-running tasks, this feature lets you stay informed without staying glued to your screen.
```bash
  pip install jupyterlab_notify[server-side-execution]
```

- **Dedicated Settings Editor**  
  A new settings editor makes customization a breeze. You can:  
  - Set the default notification mode.  
  - Adjust notification thresholds.  
  - Enable or disable Slack and email notifications.  
  - Personalize notification messages.  

  ![Settings Editor](https://github.com/user-attachments/assets/848969cd-7dc6-425f-b8a9-00eca24211fd)

- **Simplified Configuration**  
Configure notifications by creating a `jupyter_notify_config.json` file in your Jupyter config directory (e.g., `~/.jupyter/`). Check out the detailed setup guide in the README for a smooth experience. Here’s a sample configuration for Slack and email:
```json
{
  "NotificationConfig": {
    "email": "your-email@example.com",
    "slack_token": "xoxb-your-slack-bot-token",
    "slack_channel_name": "your-channel-name"
  }
}
```

### Demo: Custom Timeout Notification  
See the `custom-timeout` mode in action:

https://github.com/user-attachments/assets/fdc8f0ee-21fc-43b9-8f84-bd1ef6b957bb

### Documentation  
Dive into the updated README for a full rundown of the new features, or explore them hands-on in the refreshed [Binder notebook](https://mybinder.org/v2/gh/Darshan808/jupyterlab-notify.git/v2-test?urlpath=lab%2Ftree%2Fnotebooks%2FNotify.ipynb)

### Testing  
To ensure everything runs smoothly, this version includes:  
- Unit tests  
- Integration tests  
- Functional tests

krassowski committed:
Initial work on settings and icon in cell toolbar

krassowski committed:
Initial work on settings and icon in cell toolbar

krassowski committed:
More work on toolbar integration

krassowski committed:
More work on toolbar integration

Darshan808 committed:
explore v2 design

Darshan808 committed:
add notification feature

Darshan808 committed:
add email and slack options

Darshan808 committed:
add settings and smtp config

Darshan808 committed:
add-metadata-to-new-cells

Darshan808 committed:
add-threshold-notification

Darshan808 committed:
improve-frontend

Darshan808 committed:
improve-backend

Darshan808 committed:
add-server-tests

Darshan808 committed:
add-test-dependencies

Darshan808 committed:
add-playwright-tests

Darshan808 committed:
add-workflows-and-pre-commit

Darshan808 committed:
try-fixing-ci

Darshan808 committed:
try-ui-test-fix

Darshan808 committed:
try-fix-build-ci

Darshan808 committed:
update-eslintignore

Darshan808 committed:
modify-ui-test-workflow

Darshan808 committed:
try-fix-ui-workflow

Darshan808 committed:
modify-ui-test-workflow

Darshan808 committed:
try-fix-workflow

Darshan808 committed:
bump: upload-artifact to v4

Darshan808 committed:
Merge branch 'integrate-with-ui' into v2-test

Darshan808 committed:
wait-for-notification

Darshan808 committed:
update-readme

Darshan808 committed:
add-dummy-token

Darshan808 committed:
fix-readme

Darshan808 committed:
lint-fix

Darshan808 committed:
modify-ui-and-default-notify

Darshan808 committed:
handle-record-timing-settings

Darshan808 committed:
modify-workflow

Darshan808 committed:
force-record-timing

Darshan808 committed:
fix-workflow

Darshan808 committed:
reuse-existing-server

Darshan808 committed:
make-config-configurable

Darshan808 committed:
update-readme

Darshan808 committed:
modify-test

Darshan808 committed:
remove-bitwise-OR-use-Optional

Darshan808 committed:
add-demo-notebook-and-other-fixes

Darshan808 committed:
fix-lint

Darshan808 committed:
fix-lint-and-tests

Darshan808 committed:
fix-links

Darshan808 committed:
add-docs-images

Darshan808 committed:
fix-links

Darshan808 committed:
enhance-readme

Darshan808 committed:
remove-custom-timout-from-settings

Darshan808 committed:
add-error-msg-for-smtp-server

divyansshhh commented on .jupyter_ystore.db:
I don't think this should be needed?

divyansshhh commented on jupyterlab_notify/_version.py:
```diff
@@ -1,4 +1,4 @@
 # This file is auto-generated by Hatchling. As such, do not:
 #   - modify
 #   - track in version control e.g. be sure to add to .gitignore
-__version__ = VERSION = '2.0.0'
```
Let's add this to .gitignore

divyansshhh reviewed (commented):
Generally looks good to me, added a few minor comments. I'll install and play with this today and get back.

Darshan808 committed:
remove-unwanted-files-from-git

divyansshhh commented:
I installed and played with it, works well! I think it's good to merge if nothing is pending here cc @mlucool

singharpit94 commented:
I also spent sometime to play around with the v2 implmentation and overall I am impressed and very excited to use the new UI and integration. I have few minor feedback points which can be checked.

- I did not like how we take input in the Custom Timeout Editor, I expected to work when I entered just 2 and later realized it requires 2s like having the unit with the value, probably we can fix a default unit and user can just input the value.
-  It would be nice to have timeout settings per notebook instead of global setting.
- Do we want to batch the notification in case we have multiple notifications instead of stacking them all in the UI.
- Probably in future we may extend allow notifying users if a cell/notebook uses excessive RAM/CPU, not just time.

krassowski commented:
> I did not like how we take input in the Custom Timeout Editor, I expected to work when I entered just 2 and later realized it requires 2s like having the unit with the value, probably we can fix a default unit and user can just input the value.

Would minutes be a good default unit? We could have a dropdown for units. Alternatively, we could improve the UX by showing the validation error as-you-type.

> It would be nice to have timeout settings per notebook instead of global setting.

I agree, do you think this should be done as part of this PR or in a separate PR?

> Do we want to batch the notification in case we have multiple notifications instead of stacking them all in the UI.

I thought about it too, +1

singharpit94 commented:
> Would minutes be a good default unit? We could have a dropdown for units. Alternatively, we could improve the UX by showing the validation error as-you-type.

Yeah simplest could be to show a validation error though I think assuming seconds as the default unit should be good too

> I agree, do you think this should be done as part of this PR or in a separate PR?

Not really, probably in next version

Darshan808 committed:
add-batch-notify

Darshan808 commented:
- The default time unit for `custom-timeout` is set to seconds. 
- A batch notification timing window of 3 seconds is added. This means that after a notification is sent, any subsequent notifications received within the next 3 seconds will be grouped and sent together as a batch.

Does a 3-second window work well, or should we make this timing configurable through the settings?

singharpit94 commented:
> Does a 3-second window work well, or should we make this timing configurable through the settings?

3 second should be good IMO, having a setting is nice to have settings

> * A batch notification timing window of 3 seconds is added. This means that after a notification is sent, any subsequent notifications received within the next 3 seconds will be grouped and sent together as a batch.

Are we batching per notebook, like what happens when notification pops up from different notebooks within the 3 second time frame?

Darshan808 commented:
> Are we batching per notebook, like what happens when notification pops up from different notebooks within the 3 second time frame?

Not at the moment — notifications from all notebooks are currently batched together within the 3-second window. Would we prefer to batch them separately per notebook instead?

singharpit94 commented:
> Not at the moment — notifications from all notebooks are currently batched together within the 3-second window. Would we prefer to batch them separately per notebook instead?

Does clicking on the notification takes the user to the notebook cell, if yes batching across notebooks will be confusing on where to redirect after click?

In general we had a user request to support taking to the notebook cell on clicking on the notification.

singharpit94 commented:
On another note, I was thinking to check if v2 will still support the magics for backward compatibility or all users will need to learn the new way of setting the notifications @Darshan808

Darshan808 commented:
> On another note, I was thinking to check if v2 will still support the magics for backward compatibility or all users will need to learn the new way of setting the notifications @Darshan808

We actually removed magics completely in V2. So yes, I think users will need to adopt the new way of setting notifications.

krassowski commented:
Also, is would not be possible to install both the old and new version side-by-side (unless we rename the package).

Alternatively, we could restore the magics if this is desirable.

Darshan808 commented:
> Does clicking on the notification takes the user to the notebook cell

This isn't currently supported. However, if we want to include this feature in this PR, I’m happy to implement it.

Darshan808 committed:
restore-magics

Darshan808 committed:
add-click-to-cell-on-notification

Darshan808 committed:
bring-settings-to-notebook-level

Darshan808 commented:
### 1. **Magics Restored**
Magic commands are now supported again, so the extension can be used just like in v1 with magics.

### 2. **Clicking on Notification Navigates to Notebook Cell**
- Notifications are batched per type and per notebook. Clicking on a batched notification navigates to the **first** cell included in the batch.

### 3. **Notebook-Level Settings**
Settings have been moved to the notebook level.
I'd appreciate a review on the **UI**, especially the new notebook toolbar items I’ve added.

 **1. Notification Type Dropdown**
   A dropdown on the notebook toolbar lets users select a default notification type. All newly created cells in that notebook will use the selected type.
   New notebooks inherit the global default, which can still be changed from the Settings Editor.

**2.   Notification Threshold & Timeout Icons**
   Two icons (a bell and a bell with a clock) are added to the rightmost side of the toolbar. Clicking them opens dialogs where users can set:
   Default notification threshold
   Default timeout for custom timeout notifications

   These apply to newly created cells. New notebooks get these defaults from global settings.
   
<img width="1120" height="227" alt="image" src="https://github.com/user-attachments/assets/49a62d82-26d0-4792-bf59-90ec31f5ee08" />

I've added a video to show our new toolbar items and the "click to cell" notification feature.

https://github.com/user-attachments/assets/cc87a40c-b13a-48b6-bc50-00495ab390c7

Let me know if you have any suggestions on icons, UI or anything!

Darshan808 committed:
add-tooltips-and-notebook-toolbar-menu

Darshan808 committed:
modify to react widget

Darshan808 committed:
fix-tests

Darshan808 committed:
fix-test

Darshan808 committed:
fix-test

Darshan808 committed:
improve-icons

krassowski commented:
@singharpit94 @divyansshhh this is ready for another round of reviews :)

Darshan808 committed:
handle-null-thresholds

Darshan808 committed:
fix-lint

Darshan808 committed:
fix-binder-and-some-code

Darshan808 commented:
Hi, I had a small concern regarding UX of changing thresholds.

When a user selects the default or timeout notification type, the cell metadata stores the threshold value that is currently set at the notebook level. This allows different cells to have different threshold values, since each cell stores its own value in metadata.

However, If a user wants to change the threshold for a specific existing cell, they must first change the notebook’s default threshold and then re-apply the notification type so the cell metadata gets updated.
If the user intends to use a new threshold for a new cell but creates the cell before updating the notebook default, the old threshold is stored instead.

Does this seem like reasonable UX ?
If this seems unintuitive, one alternative would be to have buttons to change the thresholds in cell toolbar menu just like we have in notebook toolbar menu. Or are there better alternatives we should consider?

https://github.com/user-attachments/assets/91460518-13e6-4e8d-b269-ff92743e5318

CC: @singharpit94 @krassowski

Darshan808 committed:
fix-bug-with-custom-timeout-always-triggering-notification

krassowski commented:
I think that maybe error should also emit a notification when kernel dies for whatever reason, for testing purposes this can be simulated with:

```python
import os, signal
os.kill(os.getpid(), signal.SIGKILL)
```

What user ultimately cares about is whether they need to attend to the notebook.

Maybe we could observe `cell.model.executionState` as in https://github.com/deshaw/jupyterlab-execute-time/pull/138 (but listening to an appropriate signal rather than just checking the value).

singharpit94 commented:
> I think that maybe error should also emit a notification when kernel dies for whatever reason, for testing purposes this can be simulated with:
> 
> ```python
> import os, signal
> os.kill(os.getpid(), signal.SIGKILL)
> ```
> 
> What user ultimately cares about is whether they need to attend to the notebook.
> 
> Maybe we could observe `cell.model.executionState` as in [deshaw/jupyterlab-execute-time#138](https://github.com/deshaw/jupyterlab-execute-time/pull/138) (but listening to an appropriate signal rather than just checking the value).

Seems reasonable to me to add notifications for kernel erroring out.

Darshan808 committed:
notify-on-kernel-death-and-autorestarts

Darshan808 committed:
fix-lint

Darshan808 committed:
fix-test

krassowski commented on pyproject.toml:
```diff
@@ -22,10 +22,23 @@ classifiers = [
 dependencies = [
-    "jupyter_server>=2.0.1,<3"
+    "jupyter_server>=2.0.1,<3",
+    "slack_sdk>=3.35.0"
```
This is duplicated with the one in `[test]` group. Maybe add optional `[slack]` group?

Darshan808 committed:
fix-test

Darshan808 committed:
add-optional-slack-group

Darshan808 committed:
add-notebook-name-in-notification

Darshan808 committed:
lint code

Darshan808 committed:
persist-threshold

Darshan808 committed:
use-unique-keys-for-threshold-and-timeout

Darshan808 committed:
use-checkboxes-to-replace-all-thresholds

Darshan808 committed:
make-slack-and-email-messages-better

Darshan808 committed:
send-timeout-notifications-only-once

Darshan808 committed:
subclass-menu

Darshan808 committed:
seperate-tokens

Darshan808 committed:
clean code

Darshan808 committed:
fix-payload

Darshan808 committed:
do-not-display-exec-count-on-sse

Darshan808 committed:
run-lint

Darshan808 committed:
use-optional-instead

Darshan808 committed:
add-notebook-tests

Darshan808 committed:
UX: add-checkmarks

Darshan808 committed:
try-fix-test

Darshan808 committed:
try-fix-test-again

Darshan808 committed:
add-missing-wait

Darshan808 committed:
forbid-explicit-any

Darshan808 committed:
remove-debugging-leftovers

Darshan808 committed:
apply-feedbacks

Darshan808 committed:
remove-toolbar-buttons

Darshan808 committed:
use-menu-in-tests-to-change-threshold

Darshan808 committed:
run-precommit

Darshan808 committed:
address-copilot-feedbacks

Darshan808 committed:
address-feedbacks

Darshan808 committed:
Fix-lint

Darshan808 committed:
improve-notification

Darshan808 committed:
add-alwaysNotifyOnError

Darshan808 committed:
update-notebook

Darshan808 committed:
modify-docs-snapshots

Darshan808 committed:
improve-description

Darshan808 committed:
minor-comment

Darshan808 committed:
fix-kernel-death-notification

Darshan808 committed:
apply-suggestions

Darshan808 committed:
add-cell-number-for-timeout

Darshan808 committed:
run-lint

Darshan808 committed:
add-tests

Darshan808 committed:
lint-code

singharpit94 commented:
As discussed offline with @krassowski we have addressed all open feedback items hence merging this PR to master to release a rc or similar.
